src/ir/auth_logic/ast.h

```diff
@@ -152,6 +152,43 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class constantType { NumberType, PrincipalType };
+
+// Relation declaration have types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class RelationDeclarationType {
+ public:
+  using RelationDeclarationTypeVariantType =
+      std::variant<constantType, std::string>;
+  explicit RelationDeclarationType(RelationDeclarationTypeVariantType value)
+      : value_(std::move(value)) {}
+  const RelationDeclarationTypeVariantType& GetValue() const { return value_; }
+
+ private:
+  RelationDeclarationTypeVariantType value_;
+};
+
+class RelationDeclaration {
+ public:
+  explicit RelationDeclaration(std::string predicate_name, bool is_attribute,
+                               std::vector<RelationDeclarationType> arg_type)
+      : predicate_name_(std::move(predicate_name)),
+        is_attribute_(std::move(is_attribute)),
+        arg_type_(std::move(arg_type)) {}
+  const std::string& predicate_name() const { return predicate_name_; }
+  const bool& is_attribute() const { return is_attribute_; }
+  const std::vector<RelationDeclarationType>& arg_type() const {
+    return arg_type_;
+  }
+
+ private:
+  std::string predicate_name_;
+  bool is_attribute_;
+  std::vector<RelationDeclarationType> arg_type_;
```

Comment:
It is odd to call a field containing multiple argument types in a vector `arg_type_` as if it were singular. Perhaps this should be rewritten to `argument_types_`.